pipeline_steps/common.groovy

```diff
@@ -1969,3 +1974,126 @@ void updateStringParam(String name, String value, String description){
     ParametersAction newAction = existingAction.createUpdated(updatedParam)
     currentBuild.rawBuild.addOrReplaceAction(newAction)
 }
+
+/**
+* Get the predicted duration for the current build in milliseconds.
+* Prediction is based on up to 6 recent builds.
+* See: https://javadoc.jenkins-ci.org/hudson/model/Job.html#getEstimatedDurationCandidates--
+* NonCPS because build is not serialisable
+*/
+@NonCPS
+Duration getPredictedDuration(){
+  Run build = currentBuild.rawBuild
+  Job job = build.getParent()
+  return Duration.ofMillis(job.getEstimatedDuration())
+}
+
+/**
+* Get the predicted end time for this build, based on the time it moved out of the queue
+* and its predicted duration.
+* NonCPS because build is not serialisable
+*/
+@NonCPS
+LocalDateTime getPredictedCompletionTime(){
+  Run build = currentBuild.rawBuild
+  Long startTimeInSeconds = (build.getStartTimeInMillis()/1000).round(MathContext.DECIMAL32).longValueExact()
+  LocalDateTime startTime = LocalDateTime.ofEpochSecond(startTimeInSeconds, 0, ZoneOffset.UTC)
+  return startTime.plus(getPredictedDuration())
+}
+
+/**
+* Get date object representing the start of the next maintenance window
+*/
+LocalDateTime getNextMaintenanceWindowStart(){
+  // set start of maintenance window
+
+  DayOfWeek maintDay = DayOfWeek.MONDAY
+  Integer maintHour = 10
+  Integer daysInWeek = 7
+
+  // get current time & date
+  LocalDateTime now = LocalDateTime.now()
+  LocalDateTime maintStart
+
+  if (now.getDayOfWeek() == maintDay){
+    // its Monday, we either need to return today's maintenance window
+    // or next Monday's.
+    if (now.getHour() > maintHour){
+      // return next week's maintenance window
+      maintStart = now.plusDays(7)
+    } else {
+      maintStart = now
+    }
+  } else {
+    // Its not Monday.
+    Integer doW = now.getDayOfWeek().getValue()
+    Integer mDoW = maintDay.getValue()
+    Integer daysTillNextMaint = (daysInWeek -(doW - mDoW)) % daysInWeek
+    maintStart = now.plusDays(daysTillNextMaint)
+  }
+
+  // Maintstart is now a date time with the correct date, fix the time and return it.
+  return maintStart.withHour(maintHour)\
+                   .withMinute(0)\
+                   .withSecond(0)\
+                   .withNano(0)
+
+}
+
+/**
+* Check if this job is likely to run into the next maintenance window
+*/
+Boolean willOverlapMaintenanceWindow(){
+  // The pr buffer is an hour before the maintenance window starts.
+  // PRs must be predicted to finish before the PR buffer starts
+  // to be allowed to run. This gives PR jobs an hour to overrun
+  // their predicted duration.
+  Duration prBuffer = Duration.ofHours(1)
+
+  LocalDateTime completion = getPredictedCompletionTime()
+  LocalDateTime maintWindowStart = getNextMaintenanceWindowStart()
+  LocalDateTime prBufferStart = maintWindowStart.minus(prBuffer)
+
+  return prBufferStart.isBefore(completion)
+}
+
+void abortIfMaintenanceWindowPending(){
+  if (willOverlapMaintenanceWindow()){
```

Comment:
I would like to avoid scheduled jobs being aborted, as we want to increase reliability and trust. Thats my reasoning for linting the schedules. 

However I do like the idea of using an issue to signify that we have something to do in a future maintenance window. Then we can avoid aborting PR jobs if theres nothing to do in a particular week.

Reply:
Added `issueExistsForNextMaintenanceWindow()` and will only abort if that is true, great suggestion thanks. Also added a job for creating the maintenance window that has no triggers so we can run it manually when we need a window.